pkg/nimble/autoadv/shell.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2021 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     pkg_nimble_autoadv
+ * @{
+ *
+ * @file
+ * @brief       Auto advertisement module shell commands
+ *
+ * @author      Francisco Molina <francois-xavier.molina@inria.fr>
+ *
+ * @}
+ */
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "shell.h"
+#include "shell_commands.h"
+#include "xfa.h"
+
+#include "nimble_riot.h"
+#include "host/ble_hs.h"
+#include "host/ble_gap.h"
+#include "net/bluetil/ad.h"
+#include "nimble_autoadv.h"
+
+static int _ble_gap_adv_active(void)
+{
+#if MYNEWT_VAL_BLE_EXT_ADV
+    return ble_gap_ext_adv_active(CONFIG_NIMBLE_AUTOADV_INSTANCE);
+#else
+    return ble_gap_adv_active();
+#endif
+}
+
+static void _print_usage(void)
+{
+    puts("Usage:");
+    puts("\tautoadv start: start NimBLE auto advertisement");
+    puts("\tautoadv stop: stop NimBLE auto advertisement");
+    puts("\tautoadv status: print NimBLE auto advertisement status");
+}
+
+static int _autoadv_handler(int argc, char **argv)
+{
+    if (argc < 2) {
+        _print_usage();
+        return -1;
+    }
+
+    if (!strcmp(argv[1], "start")) {
+        ble_addr_t addr = { .type = nimble_riot_own_addr_type };
+        nimble_autoadv_start(&addr);
+        printf("[autoadv] shell %d: start\n", CONFIG_NIMBLE_AUTOADV_INSTANCE);
+        return 0;
+    }
+
+    if (!strcmp(argv[1], "stop")) {
+        nimble_autoadv_stop();
+        printf("[autoadv] shell %d: stop\n", CONFIG_NIMBLE_AUTOADV_INSTANCE);
+        return 0;
+    }
+
+    if (!strcmp(argv[1], "status")) {
+        if (_ble_gap_adv_active()){
+            printf("[autoadv] shell %d: active\n", CONFIG_NIMBLE_AUTOADV_INSTANCE);
+        }
+        else{
+            printf("[autoadv] shell %d: inactive\n", CONFIG_NIMBLE_AUTOADV_INSTANCE);
+        }
+        return 0;
+    }
+
+    _print_usage();
+    return -1;
+}
+
+SHELL_COMMAND(autoadv, "NimBLE autoadv", _autoadv_handler);
```

Comment:
I declared to use module `nimble_autoadv_shell` but the command is not available on the shell. Do I miss something?

Reply:
Hmmm `shell_commands` and `nimcle_autoadv` need to be used as well. But it shows upo for me:

```
USEMODULE="nimble_autoadv_shell nimble_autoadv" BOARD=dwm1001 make -C examples/gnrc_networking flash cleanterm
```
```
help
Command              Description
---------------------------------------
reboot               Reboot the node
version              Prints current RIOT_VERSION
pm                   interact with layered PM subsystem
ps                   Prints information about running threads.
ping6                Ping via ICMPv6
ping                 Alias for ping6
nib                  Configure neighbor information base
ifconfig             Configure network interfaces
rpl                  rpl configuration tool ('rpl help' for more information)
6ctx                 6LoWPAN context configuration tool
udp                  send data over UDP and listen on UDP ports
ble                  Manage BLE connections for NimBLE
autoadv              NimBLE autoadv
>
```

Reply:
Ah now, I didn't declare `nimble_autoadv`. Do you mind to pull in `nimble_autoadv` when using `nimble_autoadv_shell`?

`autoadv stop` and `autoadv status` work good, but for `autoadv start` I get:

```
2022-04-25 17:09:20,029 # autoadv start
2022-04-25 17:09:20,029 # 0x3c3
2022-04-25 17:09:20,031 # *** RIOT kernel panic:
2022-04-25 17:09:20,033 # FAILED ASSERTION.
2022-04-25 17:09:20,033 # 
2022-04-25 17:09:20,042 # 	pid | name                 | state    Q | pri | stack  ( used) ( free) | base addr  | current     
2022-04-25 17:09:20,050 # 	  - | isr_stack            | -        - |   - |    512 (  280) (  232) | 0x20000000 | 0x200001c8
2022-04-25 17:09:20,059 # 	  1 | main                 | running  Q |   7 |   1536 (  780) (  756) | 0x20000398 | 0x200006e4 
2022-04-25 17:09:20,068 # 	  2 | pktdump              | bl rx    _ |   6 |   1536 (  240) ( 1296) | 0x20002ad4 | 0x20002fe4 
2022-04-25 17:09:20,077 # 	  3 | 6lo                  | bl rx    _ |   3 |   1024 (  420) (  604) | 0x200038a8 | 0x20003b1c 
2022-04-25 17:09:20,086 # 	  4 | ipv6                 | bl rx    _ |   4 |   1024 (  452) (  572) | 0x20000a90 | 0x20000d04 
2022-04-25 17:09:20,094 # 	  5 | udp                  | bl rx    _ |   5 |   1024 (  252) (  772) | 0x20003ca8 | 0x20003fac 
2022-04-25 17:09:20,103 # 	  6 | nimble_host          | bl anyfl _ |   1 |   1024 (  372) (  652) | 0x20008ee8 | 0x20009214 
2022-04-25 17:09:20,112 # 	  7 | nimble_ctrl          | bl anyfl _ |   0 |   1024 (  248) (  776) | 0x20008ae8 | 0x20008e2c 
2022-04-25 17:09:20,121 # 	  8 | nimble_netif         | bl rx    _ |   2 |   1024 (  580) (  444) | 0x20005d84 | 0x20005fbc 
2022-04-25 17:09:20,130 # 	  9 | RPL                  | bl rx    _ |   5 |   1024 (  204) (  820) | 0x20003128 | 0x2000345c 
2022-04-25 17:09:20,136 # 	    | SUM                  |            |     |  10752 ( 3828) ( 6924)
2022-04-25 17:09:20,136 # 
2022-04-25 17:09:20,137 # *** halted.
2022-04-25 17:09:20,137 # 
2022-04-25 17:09:20,137 # 
2022-04-25 17:09:20,140 # Context before hardfault:
2022-04-25 17:09:20,141 #    r0: 0x0000000a
2022-04-25 17:09:20,143 #    r1: 0x00000000
2022-04-25 17:09:20,144 #    r2: 0x00000000
2022-04-25 17:09:20,146 #    r3: 0x00000000
2022-04-25 17:09:20,148 #   r12: 0x0000000a
2022-04-25 17:09:20,149 #    lr: 0x0000045f
2022-04-25 17:09:20,151 #    pc: 0x00000eac
2022-04-25 17:09:20,152 #   psr: 0x01000000
2022-04-25 17:09:20,152 # 
2022-04-25 17:09:20,153 # FSR/FAR:
2022-04-25 17:09:20,155 #  CFSR: 0x00000000
2022-04-25 17:09:20,156 #  HFSR: 0x80000000
2022-04-25 17:09:20,158 #  DFSR: 0x00000002
2022-04-25 17:09:20,160 #  AFSR: 0x00000000
2022-04-25 17:09:20,160 # Misc
2022-04-25 17:09:20,162 # EXC_RET: 0xffffffed
2022-04-25 17:09:20,164 # Active thread: 1 "main"
2022-04-25 17:09:20,168 # Attempting to reconstruct state for debugging...
2022-04-25 17:09:20,169 # In GDB:
2022-04-25 17:09:20,170 #   set $pc=0xeac
2022-04-25 17:09:20,171 #   frame 0
2022-04-25 17:09:20,172 #   bt
2022-04-25 17:09:20,172 # 
2022-04-25 17:09:20,175 # ISR stack overflowed by at least 16 bytes.
```

Reply:
Perfect, works for me now. Tiny nitpick: Since the argument `addr` is optional it should be `[addr]` and not `<addr>`.

Please squash :)